Sources/Web3Core/EthereumABI/ABIElements.swift

```diff
@@ -211,13 +211,67 @@ extension ABI.Element.Function {
     }
 }
 
-// MARK: - Event logs decoding
+// MARK: - Event logs decoding & encoding
 
 extension ABI.Element.Event {
     public func decodeReturnedLogs(eventLogTopics: [Data], eventLogData: Data) -> [String: Any]? {
         guard let eventContent = ABIDecoder.decodeLog(event: self, eventLogTopics: eventLogTopics, eventLogData: eventLogData) else { return nil }
         return eventContent
     }
+
+    func encodeTopic(input: ABI.Element.Event.Input, value: Any) -> EventFilterParameters.Topic? {
+        if case .string = input.type {
+            guard let string = value as? String else {
+                return nil
+            }
+            return .string(string.sha3(.keccak256).addHexPrefix())
+        } else if case .dynamicBytes = input.type {
+            guard let data = ABIEncoder.convertToData(value) else {
+                return nil
+            }
+            return .string(data.sha3(.keccak256).toHexString().addHexPrefix())
+        } else if case .address = input.type {
+            guard let encoded = ABIEncoder.encode(types: [input.type], values: [value]) else {
+                return nil
+            }
+            return .string(encoded.toHexString().addHexPrefix())
+        }
+        guard let data = ABIEncoder.convertToData(value)?.setLengthLeft(32) else {
+            return nil
+        }
+        return .string(data.toHexString().addHexPrefix())
+    }
+
+    public func encodeParameters(_ parameters: [Any?]) -> [EventFilterParameters.Topic?] {
+        guard parameters.count <= inputs.count else {
+            // too many arguments for fragment
+            return []
+        }
+        var topics: [EventFilterParameters.Topic?] = []
+
+        if !anonymous {
+            topics.append(.string(topic.toHexString().addHexPrefix()))
+        }
+
+        for (i, p) in parameters.enumerated() {
+            let input = inputs[i]
+            if !input.indexed {
+                // cannot filter non-indexed parameters; must be null
+                return []
+            }
+            if p == nil {
+                topics.append(.string(nil))
+            } else if input.type.isArray {
+                // filtering with tuples or arrays not supported
+                return []
```

Comment:
Suggestion: move it up and potentially combine into a guard statement with `input.indexed`.
```
guard input.indexed, !input.type.isArray else { return [] }
```

Reply:
This will change result when `indexed == false` && `value == nil`, I don't know if that is the right way

Reply:
I figured it out. If we have a event with 3 fields like
```
event TestEvent(string indexed f1, int[] indexed f2, int32 indexed f3);
```
When we want to build a filter to query this event with first and third fields, we have to pass `nil` to second field, and the topicFilter will be like
```
[
  '0x07c9b1ca6e62a19a2af4d11a151945afc7d47c89c716a74973b49572c1a715f4',
  '0xab036729af8b8f9b610af4e11b14fa30c348f40c2c230cce92ef6ef37726fee7',
  null,
  '0x0000000000000000000000000000000000000000000000000000000000000064'
]
```
If combine `input.indexed` and `input.type.isArray` together, we can only build a filter with first one field.